dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatcherManager.java

```diff
@@ -0,0 +1,174 @@
+            if (workerGroupTaskDispatcher.checkCloseDispatchWorkerComplete()) {
+                dispatchWorkerMap.remove(workerGroup);
+                return false;
+            }
+            return workerGroupTaskDispatcher.addTaskToWorkerGroupQueue(taskExecutionRunnable, delayTimeMills);
+        } else {
+            log.error("workerGroupTaskDispatcher {} not found, will set task {} fail",
+                    workerGroup, taskExecutionRunnable.getTaskInstance().getId());
+        }
+        return false;
+    }
+
+    /**
+     * Stops a specific worker group's task dispatch waiting queue looper.
+     *
+     * @param workerGroup the identifier for the worker group
+     */
+    private synchronized void deleteWorkerGroup(String workerGroup) {
+        WorkerGroupTaskDispatcher workerGroupTaskDispatcher = dispatchWorkerMap.get(workerGroup);
+        if (workerGroupTaskDispatcher != null) {
+            workerGroupTaskDispatcher.markDispatcherClosing();
+        } else {
+            log.warn("workerGroupTaskDispatcher {} not found", workerGroup);
+        }
+    }
+
+    /**
+     * add workerGroup
+     *
+     * @param workerGroup the identifier for the worker group
+     */
+    private synchronized void addWorkerGroup(String workerGroup) {
+        WorkerGroupTaskDispatcher workerGroupTaskDispatcher = dispatchWorkerMap.get(workerGroup);
+        if (workerGroupTaskDispatcher == null) {
+            workerGroupTaskDispatcher = new WorkerGroupTaskDispatcher(workerGroup, taskExecutorClient);
+            dispatchWorkerMap.put(workerGroup, workerGroupTaskDispatcher);
+            workerGroupTaskDispatcher.start();
+        } else {
+            workerGroupTaskDispatcher.markDispatcherStart();
+        }
+    }
+
+    /**
+     * Stop all workerGroupTaskDispatchWaitingQueueLooper
+     */
+    @Override
+    public void close() throws Exception {
+        log.info("WorkerGroupTaskDispatcherManager start close");
+        for (Map.Entry<String, WorkerGroupTaskDispatcher> entry : dispatchWorkerMap.entrySet()) {
+            try {
+                entry.getValue().markDispatcherClosing();
+            } catch (Exception e) {
+                log.error("stop worker group error", e);
+            }
+        }
+        log.info("WorkerGroupTaskDispatcherManager closed");
+    }
+
+    @Override
+    public void onWorkerGroupAdd(List<WorkerGroup> workerGroups) {
+        this.checkAndRemoveClosedDispatchWorker();
+        for (WorkerGroup workerGroup : workerGroups) {
+            this.addWorkerGroup(workerGroup.getName());
+        }
+    }
+
+    @Override
+    public void onWorkerGroupChange(List<WorkerGroup> workerGroups) {
+        this.checkAndRemoveClosedDispatchWorker();
+        String workerGroupsString = workerGroups.stream()
+                .map(WorkerGroup::getName)
+                .collect(Collectors.joining(", "));
+        log.info("Worker groups: {}", workerGroupsString);
+    }
```

Comment:
The logic is too strange, and the log is meaningless, no one can unstandard what the log means if he doesn't familiar with the code.

Reply:
There is no need to handle the change events here; just log the records. I will change it to log.info("Worker groups: {}", workerGroups);